Option bar utility

TorresCubed opened the pull request:
closes #9 
Added in functionality to option bar. 
Raised a number of states up to the window above.
slight modifications on some other areas to produce option bar usage.

TorresCubed committed:
Option Bar utility

Added theme control in the option bar.
added edit/add idea in the option bar.
header is now variable
moved modal close button to top right
changed Idea form to use text area instead

TorresCubed committed:
Option Bar utility

Added theme control in the option bar.
added edit/add idea in the option bar.
header is now variable
moved modal close button to top right
changed Idea form to use text area instead

TorresCubed commented:
I am not sure what happened, I believe I messed up my local main?
I tried to fetch from origin and rebuild what I had lost, but I think there may still be some problems...
hopefully we can work it out...

TorresCubed commented:
> Kudos, SonarCloud Quality Gate passed!
> 
> [<img alt="Bug" width="16" height="16" src="https://camo.githubusercontent.com/deaad8c9d9b09977da3f9aa455689440a17356bb359560328463022a0d9d1362/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=BUG) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=BUG)
> [<img alt="Vulnerability" width="16" height="16" src="https://camo.githubusercontent.com/60273222b26453a76e9c369287fa0106b151cbaae4fa4ce00e149b269a0fac67/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=VULNERABILITY) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=VULNERABILITY)
> [<img alt="Security Hotspot" width="16" height="16" src="https://camo.githubusercontent.com/a0723b4c04ec36cc8d010cca748afdc6300592f020130edca7403ea5b17ffe35/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742e706e67">](https://sonarcloud.io/project/security_hotspots?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=SECURITY_HOTSPOT) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/security_hotspots?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=SECURITY_HOTSPOT)
> [<img alt="Code Smell" width="16" height="16" src="https://camo.githubusercontent.com/1b640677486022b878fa0367aa2b5cffad9b009d2ff2fc3440cbd7f8da85fe55/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=CODE_SMELL) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=TorresCubed_neuromap.app&pullRequest=28&resolved=false&types=CODE_SMELL)
> 
> [<img alt="No Coverage information" width="16" height="16" src="https://camo.githubusercontent.com/eacd0d319c038e3339874e3166c6eaf3d0c647fae1a11b0f95225ea6d35302d4/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f4e6f436f766572616765496e666f2e706e67">](https://sonarcloud.io/component_measures?id=TorresCubed_neuromap.app&pullRequest=28) No Coverage information
> [<img alt="0.0%" width="16" height="16" src="https://camo.githubusercontent.com/17c8580971832aaa80226d5b1c5c61cc2f9ef6a863fc6f6bae0ab2c7d45165ed/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332e706e67">](https://sonarcloud.io/component_measures?id=TorresCubed_neuromap.app&pullRequest=28&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=TorresCubed_neuromap.app&pullRequest=28&metric=new_duplicated_lines_density&view=list)

This code smell is something that wont go away on my side... when I clear `useCallback` and then says it was defined/not used.  even though the text is gone. I have attempted to retype it a number of times.

OscarMarshall committed:
Prettified Code!

OscarMarshall committed:
Prettified Code!

OscarMarshall commented on src/Idea.js:
```diff
@@ -64,15 +70,29 @@ export const Idea = ({
+
+  let selectedStyle = selected
+    ? {
+        border: "2px solid" + ideaPackage.selectedIdeaColor,
+        boxShadow: "4px 4px 15px" + ideaPackage.selectedIdeaColor,
```
This concatenation will result in the string "2px solid#FFD700". This works only because CSS realizes what you're trying to do and fixes the mistake. These kinds of bugs are why I never use string concatenation. Instead I always prefer template strings.
```suggestion
        border: `2px solid ${ideaPackage.selectedIdeaColor}`,
        boxShadow: `4px 4px 15px ${ideaPackage.selectedIdeaColor}`,
```

OscarMarshall commented on src/IdeaForm.js:
```diff
@@ -25,13 +30,13 @@ const IdeaForm = ({ onSubmit, idea }) => {
   };
 
-  const formType = ideaState.id ? "editForm" : "addForm";
+  const formType = ideaState?.id ? "editForm" : "addForm";
```
When will `ideaState` be undefined?

TorresCubed replied:
It will not!
I think this was part of what was lost when I messed things up.

OscarMarshall commented on src/IdeaForm.js:
```diff
@@ -41,7 +46,11 @@ const IdeaForm = ({ onSubmit, idea }) => {
         <div>
-          <input type="text" value={ideaState.title} onChange={handleChange} />
+          <textarea
+            type="text"
```
`textarea` doesn't have a `type` attribute.

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,100 @@
+    (idea) => {
+      toggleFormType();
+      if (idea.title === "") return;
+      const newID = uuidv4();
```
This should go inside the following `if` statement.

OscarMarshall replied:
This still needs to be addressed.

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,100 @@
+      <div>
+        <IdeaForm
+          onSubmit={handleIdeaChange}
+          idea={ideaFormType ? selectedIdea : ""}
```
What is `ideaFormType`?

TorresCubed committed:
Small Comment Update

Cleaned some minor items, fixed code smells.

TorresCubed committed:
Small Comment Update

Cleaned some minor items, fixed code smells.

TorresCubed committed:
Prettified Code!

TorresCubed committed:
Prettified Code!

OscarMarshall commented on src/Window.js:
```diff
@@ -0,0 +1,81 @@
+import React, { useState, useReducer, useCallback } from "react";
```
```suggestion
import React, { useState, useReducer } from "react";
```
This will fix your code smell.

OscarMarshall commented on src/Window.js:
```diff
@@ -0,0 +1,81 @@
+import OptionBar from "./OptionBar";
+import update from "immutability-helper";
+
+export const IdeaContext = React.createContext({});
```
Contexts should have their own file.

TorresCubed committed:
Pull out context

Pulled context out of `Window`, separated into two files.
Established dynamic header adjustment within `FreeFormIdeas`

TorresCubed committed:
Pull out context

Pulled context out of `Window`, separated into two files.
Established dynamic header adjustment within `FreeFormIdeas`

TorresCubed committed:
Prettified Code!

TorresCubed committed:
Prettified Code!

OscarMarshall commented on src/App.css:
```diff
@@ -4,39 +4,21 @@
 .Containment {
   display: flex;
   overflow: hidden;
+  position: absolute;
```
Do we need this?

TorresCubed replied:
Nope!

OscarMarshall commented on src/App.js:
```diff
@@ -1,25 +1,22 @@
-            <Window />
-          </DndProvider>
-        </div>
+const App = () => {
+  return (
```
```suggestion
const App = () => (
```

OscarMarshall commented on src/App.js:
```diff
@@ -1,25 +1,22 @@
+      <header className="App-header">
+        <h1>Map Your Ideas Below</h1>
+      </header>
+      <div>
```
Do we need this surrounding `div`?

TorresCubed replied:
It does not appear so

OscarMarshall commented on src/FreeFormIdeas.css:
```diff
@@ -0,0 +1,11 @@
+.FreeFormIdeas {
+  top: inherit;
```
Do we need this?

TorresCubed replied:
Not anymore now that you helped me with `getBoundingClientRectangle()`

OscarMarshall commented on src/FreeformIdeas.js:
```diff
@@ -49,85 +56,52 @@ export function calcCoords(idea, arrowRotation) {
-      },
-    }
-  );
+  const themePackage = useContext(ThemeContext);
```
```suggestion
  const { theme } = useContext(ThemeContext);
```

OscarMarshall commented on src/FreeformIdeas.js:
```diff
@@ -49,85 +56,52 @@ export function calcCoords(idea, arrowRotation) {
-  );
+  const themePackage = useContext(ThemeContext);
+
+  const ideaPackage = useContext(IdeaContext);
+  const ideas = ideaPackage.ideas;
+  const selectedId = ideaPackage.selectedId;
```
```suggestion
  const { ideas, selectedId } = useContext(IdeaContext);
```

OscarMarshall commented on src/FreeformIdeas.js:
```diff
@@ -136,52 +110,72 @@ const FreeFormIdeas = () => {
+    [drop]
+  );
+
+  useLayoutEffect(() => {
```
This can be a `useEffect` since we don't need to run it every DOM update.

OscarMarshall replied:
In fact, we don't need this at all. This should live in domElementRef.

OscarMarshall commented on src/Idea.js:
```diff
@@ -15,17 +17,19 @@ export const Idea = ({
-  ideasDispatch,
 }) => {
-  const [domElement, setDomElement] = useState();
+  const themePackage = useContext(ThemeContext);
 
+  const ideaPackage = useContext(IdeaContext);
```
Deconstruct these

OscarMarshall commented on src/Idea.js:
```diff
@@ -64,17 +68,22 @@ export const Idea = ({
       className={"idea" + (selected ? " selected" : "")}
-      onClick={select}
       onMouseUp={linkerDesignation}
+      onClick={select}
```
`handleSelect`

OscarMarshall commented on src/Modal.js:
```diff
@@ -5,7 +5,9 @@ const Modal = ({ onClose, children }) => (
       {children}
-      <button onClick={onClose}>Exit</button>
+      <div className="closeButton" onClick={onClose}>
+        X
```
```suggestion
        &times;
```

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,113 @@
+import { SketchPicker } from "react-color";
 
 const OptionBar = () => {
+  const ideaPackage = useContext(IdeaContext);
+  const selectedId = ideaPackage.selectedId;
+  const selectedIdea = ideaPackage.ideas[selectedId];
+
+  const themePackage = useContext(ThemeContext);
```
Destructure these.

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,113 @@
+  const changeColor = useCallback(
+    (color) => {
+      setBackground(color.hex);
+      switch (colorSelectorValue) {
+        case "1":
+          themePackage.updateTheme({
+            element: "optionBarColor",
+            color: color.hex,
+          });
+          break;
+        case "2":
+          themePackage.updateTheme({
+            element: "freeFormIdeasColor",
+            color: color.hex,
+          });
+          break;
+        case "3":
+          themePackage.updateTheme({
+            element: "selectedIdeaColor",
+            color: color.hex,
+          });
+          break;
+        default:
+          return;
+      }
```
The only thing that changes between all of these cases is the `element` attribute.
```suggestion
      themePackage.updateTheme({
        element: { 1: "optionBarColor", 2: "freeFormIdeasColor", 3: "selectedIdeaColor" }[colorSelectorValue],
        color: color.hex,
      });
```

TorresCubed replied:
jeez

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,113 @@
+  const themePackage = useContext(ThemeContext);
+  const [colorSelectorValue, setColorSelectorValue] = useState(0);
+  const [background, setBackground] = useState("#800080");
+  const [hiding, setHiding] = useState({ display: "none" });
```
Having possible values of `hiding` as `{ display: "none" }` and `{ display: "flex" }` is cryptic. I would make the possible values `true` or `false`.

OscarMarshall commented on src/OptionBar.js:
```diff
@@ -1,9 +1,113 @@
+          <option value="3">Selection Outline</option>
+        </select>
+      </div>
+      <div style={hiding}>
+        <SketchPicker onChange={changeColor} color={background} />
+      </div>
```
If you make `hiding` be a boolean, you can change this to be 
```suggestion
      {hiding && <SketchPicker onChange={changeColor} color={background} />)
```

OscarMarshall commented on src/Window.js:
```diff
@@ -0,0 +1,21 @@
+import { ThemeContext, ThemeStateManagementProvider } from "./ThemeContext";
+import { IdeaContext, IdeaStateManagementProvider } from "./IdeaContext";
+
+const Window = () => {
+  return (
```
```suggestion
const Window = () => (
```

TorresCubed replied:
Changing the contexts to hooks and placing them inside window, now means I need to designate what is returned, right?

OscarMarshall replied:
Yes

TorresCubed committed:
Option Bar Utility CR update

Changed the context value providers into hooks to maintain proper use.
Refactored on suggested edits

TorresCubed committed:
Option Bar Utility CR update

Changed the context value providers into hooks to maintain proper use.
Refactored on suggested edits

TorresCubed committed:
Prettified Code!

TorresCubed committed:
Prettified Code!

OscarMarshall commented on src/Idea.js:
```diff
@@ -64,17 +68,22 @@ export const Idea = ({
-      style={{ left, top }}
+      style={{ left, top, ...selectedStyle }}
       onDoubleClick={handleDoubleClick}
       className={"idea" + (selected ? " selected" : "")}
```
```suggestion
      className={`idea${selected ? " selected" : ""}`}
```

TorresCubed committed:
Merge branch 'option-Bar-Utility' of https://github.com/TorresCubed/neuromap.app into option-Bar-Utility

TorresCubed committed:
Update .gitignore
